x-pack/solutions/observability/plugins/apm/public/components/shared/trace_waterfall/use_trace_waterfall.ts

```diff
@@ -95,6 +96,51 @@ export function getTraceParentChildrenMap(traceItems: TraceItem[]) {
   return traceMap;
 }
 
+export enum TraceDataState {
+  Full,
+  Partial,
+  Empty,
+}
+
+function matchParentId(this: TraceItem, parent: TraceItem) {
+  return parent.id === this.parentId;
+}
+
+function getFallbackRootItem(item: TraceItem, _index: number, traceItems: TraceItem[]): boolean {
+  // Avoid allocating a whole new function every iteration, making this operation less expensive
+  return !traceItems.find(matchParentId, item);
+}
+
+export function getRootItemOrFallback(
+  traceParentChildrenMap: Record<string, TraceItem[]>,
+  traceItems: TraceItem[]
+) {
+  const rootItem = traceParentChildrenMap.root?.[0];
```

Comment:
I just checked the `getTraceParentChildrenMap` and it's not 100% true that a parentless transaction is the root transaction https://github.com/elastic/kibana/issues/212931.